### PR TITLE
Enable SocketAPM to run on LwIP and add DDS support

### PR DESCRIPTION
**Pull request by magicrub**
Enable LwIP support to the existing SocketAPM so all libraries that use SocketAPM should auto-magically be able to work on SMT32 targets. More work to come to expand on this...

It also adds a new a pram DDS_IP which needs some param-name discussion because DDS_PORT is wrong.

This is on top of the Networking cleanup PR https://github.com/ArduPilot/ardupilot/pull/25522

Thanks to @tridge and @Ryanf55

### CODE REVIEW

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 1]
```diff
@@ -65,6 +65,10 @@ const AP_Param::GroupInfo AP_DDS_Client::var_info[] {
     // @User: Standard
     AP_GROUPINFO("_PORT", 2, AP_DDS_Client, udp.port, 2019),
 
+    // @Group: _IP
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:68]: Is there a way to add a description
"IP address of the DDS Agent"

**Review by Ryanf55 — CHANGES_REQUESTED** [review]
Right now, it doesn't seem to work in SITL, the LWP thread exits. 

image

**Comment by magicrub** [comment]
>Right now, it doesn't seem to work in SITL, the LWP thread exits.

try `./waf configure --board sitl --enable-dds --enable-networking`

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_DDS/AP_DDS_Client.cpp:68]: I don't think so.

Note, the _OPT docs are broken. Docs say "DDS_UDP_PORT" but code is "DDS_PORT". I'm in favor of removing the "_UDP_" part and have a different param control TCP/UDP server/client options

**Comment by Ryanf55** [comment]
> > Right now, it doesn't seem to work in SITL, the LWP thread exits.
> 
> try `./waf configure --board sitl --enable-dds --enable-networking`

Thanks. It works on SITL. On Pixhawk 6X, I can't compile now:
```
$ ./waf configure --board Pixhawk6X --enable-dds --enable-networking
$  ./waf plane --upload
[  70/1155] Compiling libraries/APM_Control/AP_PitchController.cpp
In file included from ../../libraries/AP_HAL/board/chibios.h:3,
                 from ../../libraries/AP_HAL/AP_HAL_Boards.h:137,
                 from ../../libraries/AP_Scripting/AP_Scripting_config.h:3,
                 from ../../libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.cpp:1:
./hwdef.h:48: error: "AP_NETWORKING_ENABLED" redefined [-Werror]
   48 | #define AP_NETWORKING_ENABLED               TRUE
      | 
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors

In file included from ../../libraries/AP_HAL/board/chibios.h:3,
                 from ../../libraries/AP_HAL/AP_HAL_Boards.h:137,
                 from ../../libraries/AP_Math/control.h:3,
                 from ../../libraries/AC_WPNav/AC_WPNav_OA.cpp:1:
./hwdef.h:48: error: "AP_NETWORKING_ENABLED" redefined [-Werror]
   48 | #define AP_NETWORKING_ENABLED               TRUE
      | 
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors

In file included from ../../libraries/AP_HAL/board/chibios.h:3,
                 from ../../libraries/AP_HAL/AP_HAL_Boards.h:137,
                 from ../../libraries/AP_AHRS/AP_AHRS_config.h:3,
                 from ../../libraries/AP_AHRS/AP_AHRS.h:24,
                 from ../../libraries/APM_Control/AP_SteerController.cpp:20:
./hwdef.h:48: error: "AP_NETWORKING_ENABLED" redefined [-Werror]
   48 | #define AP_NETWORKING_ENABLED               TRUE
      | 
...

      | 
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors

In file included from ../../libraries/AP_HAL/board/chibios.h:3,
                 from ../../libraries/AP_HAL/AP_HAL_Boards.h:137,
                 from ../../libraries/AP_HAL/AP_HAL.h:6,
                 from ../../libraries/AP_Param/AP_Param.h:27,
                 from ../../libraries/AC_AttitudeControl/AC_AttitudeControl.h:7,
                 from ../../libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.h:6,
                 from ../../libraries/AC_AttitudeControl/AC_AttitudeControl_TS.h:6,
                 from ../../libraries/AC_AttitudeControl/AC_AttitudeControl_TS.cpp:20:
./hwdef.h:48: error: "AP_NETWORKING_ENABLED" redefined [-Werror]
   48 | #define AP_NETWORKING_ENABLED               TRUE
      | 
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
```

Is this because both chibios_hwdef.py and AP_Networking_Config.h have `AP_NETWORKING_ENABLED`, but only one is protected?

**Comment by magicrub** [comment]
It's messy at the moment. The enable-networking only applies to SITL. For chibos it's defined by the hwdef.

**Comment by Ryanf55** [comment]
> It's messy at the moment. The enable-networking only applies to SITL. For chibos it's defined by the hwdef.

Thanks. To set up xinetd:
* https://en.wikipedia.org/wiki/Xinetd
* https://github.com/xinetd-org/xinetd
* https://manpages.ubuntu.com/manpages/jammy/en/man8/xinetd.8.html

`sudo apt install xinetd`

 change the settings
```
$ cat /etc/xinetd.d/echo
# default: off
# description: An xinetd internal service which echo's characters back to
# clients.
# This is the tcp version.
service echo
{
        disable         = yes
        type            = INTERNAL
        id              = echo-stream
        socket_type     = stream
        protocol        = tcp
        user            = root
        wait            = no
}

# This is the udp version.
service echo
{
        disable         = no
        type            = INTERNAL
        id              = echo-dgram
        socket_type     = dgram
        protocol        = udp
        user            = root
        wait            = yes
}
```

Restart xinetd: 
`sudo systemctl restart xinetd.service`

Results:
```

**Comment by Ryanf55** [comment]
https://github.com/ArduPilot/ardupilot/pull/25522 is merged, please rebase this and fix conflicts.

**Comment by Ryanf55** [comment]
Next step: Add an AP_Networking_SITL as a backend. The UDP request/response doesn't work in DDS, but works with the test server. In order to have an easier time debugging, we would like to add a SITL backend.  Tridge agrees we can merge this as-is even though DDS doesn't yet work.

**Review by Ryanf55 — APPROVED** [review]

File: libraries/AP_Networking/AP_Networking_Config.h [hunk 2]
```diff
@@ -43,3 +60,13 @@
...
 #define AP_NETWORKING_DEFAULT_MAC_ADDR "C2:AF:51:03:CF:46"
 #endif
+
+#ifndef AP_NETWORKING_TESTS_ENABLED
+#define AP_NETWORKING_TESTS_ENABLED 0
+#endif
+
+#if AP_NETWORKING_TESTS_ENABLED
+#ifndef AP_NETWORKING_TEST_IP
+#define AP_NETWORKING_TEST_IP "192.168.13.15"
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking_Config.h:70]: Make this the same value as the one in here: `AP_DDS_DEFAULT_UDP_IP_ADDR "192.168.13.2"`

**Review by Ryanf55 — DISMISSED** [review]

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking_Config.h]: fixed

**Comment by magicrub** [comment]
@Ryanf55 rebased as requested.

**Comment by Ryanf55** [comment]
> @Ryanf55 rebased as requested.

Let's merge it and solve DDS later. The UDP tests work; it might work fine for mavlink as-is.

**Comment by magicrub** [comment]
I'm good with that. Mavlink support will be another can of worms that should be in a new PR

**Review by tridge — APPROVED** [review]

**Pull request merged**